include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,250 @@
+       */
+      unsigned int threshold_n_linear_iterations;
+
+      /**
+       * Reuse NOX solver instance in the next non-linear solution.
+       */
+      bool reuse_solver;
+    };
+
+    /**
+     * Constructor.
+     *
+     * If @p parameters is not filled, a Newton solver with full step is used.
+     * An overview of possible parameters is given at
+     * https://docs.trilinos.org/dev/packages/nox/doc/html/parameters.html.
+     */
+    NOXSolver(AdditionalData &                            additional_data,
+              const Teuchos::RCP<Teuchos::ParameterList> &parameters =
+                Teuchos::rcp(new Teuchos::ParameterList));
+
+    /**
+     * Clear internal state.
+     */
+    void
+    clear();
+
+    /**
+     * Solve non-linear problem and return number of iterations.
+     */
+    unsigned int
+    solve(VectorType &solution);
+
+    /**
+     * User function that computes the residual.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x, VectorType &f)> residual;
+
+    /**
+     * User function that sets up the Jacobian.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x)> setup_jacobian;
+
+    /**
+     * User function that sets up the preconditioner for inverting
+     * the Jacobian.
+     *
+     * @note The function is optional and is used when setup_jacobian is
+     * called and the preconditioner needs to be updated (see
+     * update_preconditioner_predicate and
+     * AdditionalData::threshold_nonlinear_iterations).
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x)> setup_preconditioner;
+
+    /**
+     * User function that applies the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations. For instance, this function is required if the
+     * polynomial line search (@p NOX::LineSearch::Polynomial) is
+     * chosen, whereas for the full step case (@p NOX::LineSearch::FullStep)
+     * it won't be called.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x, VectorType &v)> apply_jacobian;
+
+    /**
+     * User function that applies the inverse of the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<
+      int(const VectorType &f, VectorType &x, const double tolerance)>
+      solve_with_jacobian;
+
+    /**
+     * User function that applies the inverse of the Jacobian and
+     * returns the numer of linear iterations the linear solver needed.
+     *
+     * @note This function should return -1 in the case of failure.
```

Comment:
Just checking: Does it need to return `0` in the case of success, like the others?